st3/sublime_lib/resource_path.py

```diff
@@ -0,0 +1,83 @@
+import sublime
+import os.path
+from collections import OrderedDict
+
+from .pure_resource_path import PureResourcePath
+from .path_compat import Path
+from .glob_util import get_glob_expr
+
+
+__all__ = ['glob_resources', 'ResourcePath']
+
+
+def glob_resources(pattern):
+    expr = get_glob_expr(pattern)
+    return sorted(
+        ResourcePath(path) for path in sublime.find_resources('')
+        if expr.match(path)
+    )
+
+
+class ResourcePath(PureResourcePath):
+    def file_path(self):
+        """Return a Path object representing a filesystem path inside the
+        Packages or Cache directory. Even if there is a resource at this
+        `ResourcePath`, there may not be a file at that filesystem path.
+
+        If `pathlib` is available, return a :class:`pathlib.Path`. Otherwise,
+        return a stub object that is compatible with :func:`str` and
+        :func:`os.fspath`.
+
+        :raise ValueError: if ``ResourcePath.root`` is not `Packages` or
+        `Cache`.
+        """
+        if self.root == 'Packages':
+            return Path(os.path.join(sublime.packages_path(), *self.parts[1:]))
+        elif self.root == 'Cache':
+            return Path(os.path.join(sublime.cache_path(), *self.parts[1:]))
+        else:
+            raise ValueError("%r is not a packages or cache path" % (self,))
+
+    def exists(self):
+        """Return ``True`` if there is a resource at this path, or ``False``
+        otherwise."""
+        return str(self) in sublime.find_resources(self.name)
+
+    def read_text(self):
+        """Load the resource at this path and return it as text.
+
+        :raise FileNotFoundError: if there is no resource at this path.
+
+        :raise UnicodeDecodeError: if the resource cannot be decoded as UTF-8.
+        """
+        try:
+            return sublime.load_resource(str(self))
+        except IOError as err:
+            raise FileNotFoundError from err
+
+    def read_bytes(self):
+        """Load the resource at this path and return it as bytes.
+
+        :raise FileNotFoundError: if there is no resource at this path.
+        """
+        try:
+            return sublime.load_binary_resource(str(self))
+        except IOError as err:
+            raise FileNotFoundError from err
+
+    def glob(self, pattern):
+        base = str(self) + '/' if self._parts else ''
+        return glob_resources(base + pattern)
+
+    def rglob(self, pattern):
+        return self.glob('**/' + pattern)
+
+    def children(self):
+        depth = len(self._parts)
+        return [
+            self / next_part
+            for next_part in OrderedDict.fromkeys(
```

Comment:
Why iterate over an ordered dict created with fromkeys? Might as well just use a normal generator, which is also ordered.

```py
return [self / resource.parts[depth] for resource in self.glob('**')
```

Actually, why not just use `self.glob('*')`?

Actually-actually, why have this in the first place?

Reply:
This method gets all immediate child paths that contain a resource or are the prefix of a path that contains a resource. If this were a filesystem, it would represent files and directories, but because resources have no concept of directories, the interpretation is slightly different and the implementation is substantially different.

`OrderedDict.fromkeys` gets unique values in order.

Reply:
Ah, I get it now. `chilren` includes paths that would be directories but don't (really) exist in resource paths. I suggest we add a method that tests whether a resource path would point to a directory (by checking whether there are paths that match `str(dir_path) + '/**'`).

Reply:
Even in that case, the path could still point to a file. For example, you could have an installed (zipped) package containing the resource `/Packages/Foo/bar/baz` and an overriding file at `/Packages/Foo/bar`. Then, the resource path `/Packages/Foo/bar` maps to a file but also has descendants. In a filesystem, files and directories are exclusive categories — a path may point to one or the other but never both. Because of this, I think that calling that path a “directory” could lead users to make bad assumptions.

Reply:
Ugh, yeah, that would suck.

While we could argue that override packages always override the corresponding installed package, I really don't know how to handle this because most likely ST will *still* see all sub-resources in the installed package where it is a "directory" despite the override being a file. I've never come across this situation, but it is a possibility.

For now, let's concentrate on actionable features and leave this for later.

Reply:
Would you mind going through and resolving some of the conversations?

Reply:
I was hoping the outdated ones would get resolved automatically, but I guess they disabled that once resolving was implemented.